### PR TITLE
chore: introduce initial version of flakiness dashboard

### PR DESCRIPTION
**Pull request by aslushnikov**
This patch introduces a dashboard that records test results and
uploads them to https://github.com/aslushnikov/puppeteer-flakiness-dashboard

Since many bots might push results in parallel, each bot pushes
results to its own git branch.

FlakinessDashboard also generates a simple README.md with a flakiness
summary. If this proves to be not enough, we can build a website that
fetches flakiness data and renders it nicely.

### CODE REVIEW

File: test/test.js [hunk 1]
```diff
@@ -105,10 +105,52 @@ if (process.env.CI && testRunner.hasFocusedTestsOrSuites()) {
...
+  const {FlakinessDashboard} = require('../utils/flakiness-dashboard');
+  const sha = process.env.CIRRUS_CHANGE_IN_REPO;
+  const dashboard = await FlakinessDashboard.initialize({
+    dashboardName: process.env.CIRRUS_TASK_NAME,
+    build: {
+      url: `https://cirrus-ci.com/build/${process.env.CIRRUS_BUILD_ID}`,
+      name: sha.substring(0, 8),
+    },
+    dashboardRepo: {
+      url: 'https://github.com/aslushnikov/puppeteer-flakiness-dashboard.git',
```
Comment (JoelEinbinder) [inline test/test.js]: puppeteer org?

**Comment by JoelEinbinder** [comment]
> Since many bots might push results in parallel, each bot pushes
> results to its own git branch.

Who guarantees that even a single bot won't run two instances in parallel?

**Comment by aslushnikov** [comment]
> Who guarantees that even a single bot won't run two instances in parallel?

Added retries with cooldown in case there are several instances of the same bot working on different revisions in parallel. This, however, should not happen in our setup since CIs limit free accounts to one bot instance.

**Review by MedRedha — APPROVED** [review]

File: test/test.js [hunk 2]
```diff
@@ -111,4 +111,67 @@ new Reporter(testRunner, {
   projectFolder: utils.projectRoot(),
   showSlowTests: process.env.CI ? 5 : 0,
 });
+
+// Unconditionally generate testIDs for all tests and verify they don't clash.
+// This will add |test.testId| for every test.
+generateTestIDs(testRunner);
```
Comment (JoelEinbinder) [inline test/test.js]: nit: can we put this into its own file. I want to be able to easily remove the flakiness dashboard code if we decide its not worth it.

File: utils/flakiness-dashboard/FlakinessDashboard.js [hunk 3]
```diff
@@ -0,0 +1,276 @@
...
+
+const RED_COLOR = '\x1b[31m';
+const GREEN_COLOR = '\x1b[32m';
+const YELLOW_COLOR = '\x1b[33m';
+const RESET_COLOR = '\x1b[0m';
+
+class FlakinessDashboard {
+  /**
+   * @param {{url: string, branch: string, username: string, password: string}} dashboardRepo
+   * @param {{url: string, name: string}} checkout
```
Comment (JoelEinbinder) [inline utils/flakiness-dashboard/FlakinessDashboard.js]: what are all these other params here?

**Review by JoelEinbinder — APPROVED** [review]

File: utils/flakiness-dashboard/FlakinessDashboard.js (hunk 3 above)
Comment (aslushnikov) [reply utils/flakiness-dashboard/FlakinessDashboard.js]: Indeed

File: test/test.js (hunk 2 above)
Comment (aslushnikov) [reply test/test.js]: Moved to `//test/utils.js` and consolidated in a single function call.

**Pull request merged**